### DISCUSSION TITLE
[Miscellaneous] Optimism / OP Stack based blockchain might have been not applying L1 Fee Scalar properly in transcation fee calculations

### DISCUSSION THREAD

**Discussion by jonathanykh** [ethereum-optimism/developers#46] [Miscellaneous] [answered]
### What is the problem

While I have been looking into random transactions on Optimism Mainnet Explorer, it appeared to me that the transaction fees charged as displayed on the explorer have not taken into account of the `L1 Fee Scalar`, as described in Optimism's official help book and the block explorer to be a part of the factor in the `L1 security fee` calculation.

Per below screenshot of an [example transaction](https://optimistic.etherscan.io/tx/0x84e128b7c54d22f5b283538553cc0a9d9c8056c35a1010d64558dd0d86fc1fc3), the transaction fee is found to be calculated as:
```
Tx fee charged
= L2 gas consumed * L2 Gas Price + L1 Gas consumed * L1 Gas Price
= 21000 * 0.000000000011341535 + 1,389.88799999998 * 0.000000048852715769
= 0.00006813797565 (Matching what's shown on the screencap)
```
However, based on Optimism's official help book and the block explorer, the transaction should be:
```
Tx fee should be charged
= L2 gas consumed * L2 Gas Price + L1 Gas consumed * L1 Gas Price * L1 Fee Scalar
= 21000 * 0.000000000011341535 + 1,389.88799999998 * 0.000000048852715769 * 0.684
= 0.00004668163777 (Does not match what's shown on the screencap)
```

<img width="955" alt="image" src="https://github.com/ethereum-optimism/developers/assets/5406198/6b90ace5-b3e8-41fb-a3b4-1428a08ed043">

Therefore, the next step is to identify whether this issue is blockchain explorer's frontend display problem, which could be just a mis-configured formula, or if this is an backend/chain configuration issue where the `L1 Fee Scalar` has not been actually applied.

I took a lazy but easy to verify approach here by looking at the pre-transaction and post-transaction account balance state on Tenderly, by simulating two transactions of 1 wei from zero address sent to the user who originated the transaction above, with one simulation before and one siulation after the actual balance change, in order to see the account balance state before and after the actual transaction originated by the user.

The [first simulation](https://www.tdly.co/shared/simulation/b63d798b-f485-4528-bc55-6ed5ae4b98fc) was done at block 111710718, or one block before the transaction: 
<img width="451" alt="image" src="https://github.com/ethereum-optimism/developers/assets/5406198/9aa6d659-659e-402e-b81d-fc7384f97a30">
This screenshot indicates at block 111710718, the user had an account balance of 27910566951306927 wei

The [second simulation](https://www.tdly.co/shared/simulation/86961502-989c-4a70-8821-4a2ce8357084) was done at block 111710720, or one block after the transaction
<img width="456" alt="image" src="https://github.com/ethereum-optimism/developers/assets/5406198/048cefda-99f3-424e-96be-28a8dfe994d9">
This screenshot indicates at block 111710720, the user had an account balance of 27342428975657184 wei

While the user has sent out the transaction with 0.0005 ETH, as shown below:
<img width="945" alt="image" src="https://github.com/ethereum-optimism/developers/assets/5406198/9c61bbf2-3c02-41c8-9f6e-b57645e23667">

Therefore by negating the 0.0005 ETH spent as transaction value, the rest of the balance change should belong to gas payments, which is a formula of:
```
Gas payment (wei) = (27910566951306927 - 0.0005*1e18) - 27342428975657184 = 68137975649743 wei = 0.00006813797565 ether
```
This shows to have an exact match to the gas fee charged as displayed on the block explorer, therefore indicating the `L1 Fee Scalar` has not been applied to user's transaction gas cost calcualtions.

Tracing on the transaction illustarted above, user has also shown no additional value in or out from block 111710718 to 111710720, ruling out the possibility that the user might have been credited/debited balances from his account state.
<img width="1378" alt="image" src="https://github.com/ethereum-optimism/developers/assets/5406198/d4ae986c-4f51-4310-a4f8-9273b78cd929">
<img width="1415" alt="image" src="https://github.com/ethereum-optimism/developers/assets/5406198/cdea4d22-a7d9-4a2a-87f0-48c53c7929a6">

### Steps to reproduce

Problematic calculations could be found by looking into any transactions on OP Mainnet Explorer or Base Explorer

https://optimistic.etherscan.io/tx/0x84e128b7c54d22f5b283538553cc0a9d9c8056c35a1010d64558dd0d86fc1fc3

### Notes
Thank you Optimism team and contributors from Discord for encouraging me to share this issue here on Github.

### Additional Information
The same issue is also found on transactions took place on Base, which lead to the potential of any OP Stack chains could also be affected.

**Comment by sbvegan** [comment]
Hey @jonathanykh, thank you again for flagging this. I've raised it internally and we'll get some more eyes on it. Your math seems to check out. I'll see what other folks say.

**Comment by ajsutton** [comment, marked as answer]
I believe Etherscan is applying the L1 fee scalar to the L1 gas used.  Note that the "L1 Gas Used by Txn" is not an integer but there's no such thing as decimal gas.  If you request the transaction receipt from an actual node it gives:

```
cast receipt -j --rpc-url $OP_MAINNET 0x84e128b7c54d22f5b283538553cc0a9d9c8056c35a1010d64558dd0d86fc1fc3 | jq
{
  "transactionHash": "0x84e128b7c54d22f5b283538553cc0a9d9c8056c35a1010d64558dd0d86fc1fc3",
  "transactionIndex": "0x4",
  "blockHash": "0xe3b91307f84a3ac0fd92f86cb92468328b0597e1f5211ff7bfedefdc49b66049",
  "blockNumber": "0x6a891ff",
  "from": "0x2e086aa3c4fa88cd17a80842199c59f6597551d0",
  "to": "0xe60ea7b0725e252ce2ad9f8e79eb3e686f5d1148",
  "cumulativeGasUsed": "0x4422d",
  "gasUsed": "0x5208",
  "contractAddress": null,
  "logs": [],
  "status": "0x1",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "type": "0x2",
  "effectiveGasPrice": "0xad0edf",
  "l1Fee": "0x3dc127a9b0d7",
  "l1FeeScalar": "0.684",
  "l1GasPrice": "0xb5fd948f9",
  "l1GasUsed": "0x7f0"
}
```

That `l1GasUsed: 0x7f0` is 2032 in decimal. `2032 * 684000 / 1000000 = 1,389.888`.

**Reply by opfocus** [reply]
Do you agree that the reason for the confusion is that the l1GasUsed we have here does not have the same caliber as the l1GasUsed shown on the block explorer

**Reply by jonathanykh** [reply]
Thanks. That makes sense to me. I tried your method as well on an example [transaction](https://basescan.org/tx/0x915491f6782a0e9045c806a1de438b9647dabc8a8e7c4eddacb871e2d875bdd3) on Base, and obtained the following:

```
cast receipt -j --rpc-url https://base.llamarpc.com 0x915491f6782a0e9045c806a1de438b9647dabc8a8e7c4eddacb871e2d875bdd3 | jq
{
  "transactionHash": "0x915491f6782a0e9045c806a1de438b9647dabc8a8e7c4eddacb871e2d875bdd3",
  "transactionIndex": "0x9",
  "blockHash": "0x7aa37f0c334a38630afa87544d721ea50b7c6c85dcd83be3c89d46c72e33e2b9",
  "blockNumber": "0x5db602",
  "from": "0xfd6321939330d45269f2d3b87aaa7cc1128cc23a",
  "to": "0x2fc617e933a52713247ce25730f6695920b3befe",
  "cumulativeGasUsed": "0xe581c",
  "gasUsed": "0x5208",
  "contractAddress": null,
  "logs": [],
  "status": "0x1",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "type": "0x2",
  "effectiveGasPrice": "0x37b53",
  "l1Fee": "0xf0a31390c8b",
  "l1FeeScalar": "0.684",
  "l1GasPrice": "0x2bf9df458",
  "l1GasUsed": "0x800"
}
```
Of which "l1GasUsed": "0x800", or 2048 in decimals, equals `1400.832 / 0.684`. Which I believe we can confirm the same issue also happens on Base and likely any other OP chains with Etherscan as official explorer.

Wanted to ask if there will be any next steps I need to take on this, for example raising this issue on another repo, or if the Optimism team will take over and communicate this internally to the Etherscan team.

**Reply by Chomtana** [reply]
Notice in this receipt response

"l1Fee": "0x3dc127a9b0d7",

Which means 0.000067899803414743 ETH

So, it means that the problem comes from op-geth json rpc response.

**Reply by ajsutton** [reply]
From that receipt:
```
  "l1Fee": "0xf0a31390c8b",
```
not `0x3dc127a9b0d7`. I think you've copied the value from the op-mainnet example and are comparing it to the base example. :)

**Reply by Chomtana** [reply]
![image](https://github.com/ethereum-optimism/developers/assets/4103490/89039fc7-2bb7-418b-81ae-9c7c2a1b5a64)

<img width="508" alt="image" src="https://github.com/ethereum-optimism/developers/assets/4103490/e24ad2f6-94a8-4e11-8bc9-c08d96e0888d">

Is this two related?

**Reply by ajsutton** [reply]
I don't know what transaction that screenshot is from.  Can you provide the actual link and copy of the cast output instead of screenshots please? It makes it much easier to see read and work with the data.

**Reply by Chomtana** [reply]
So, L1 Gas Used by Txn in etherscan = L1 Gas Used * L1 Fee Scalar
If this is true everything seem to be correct.

**Reply by Chomtana** [reply]
Tx: https://optimistic.etherscan.io/tx/0x84e128b7c54d22f5b283538553cc0a9d9c8056c35a1010d64558dd0d86fc1fc3

Request:
```json
{
	"jsonrpc":"2.0",
	"method":"eth_getTransactionReceipt",
	"params":[
		"0x84e128b7c54d22f5b283538553cc0a9d9c8056c35a1010d64558dd0d86fc1fc3"
	],
	"id":1
}
```

Response:
```json
{
    "jsonrpc": "2.0",
    "result": {
        "blockHash": "0xe3b91307f84a3ac0fd92f86cb92468328b0597e1f5211ff7bfedefdc49b66049",
        "blockNumber": "0x6a891ff",
        "contractAddress": null,
        "cumulativeGasUsed": "0x4422d",
        "effectiveGasPrice": "0xad0edf",
        "from": "0x2e086aa3c4fa88cd17a80842199c59f6597551d0",
        "gasUsed": "0x5208",
        "l1Fee": "0x3dc127a9b0d7",
        "l1FeeScalar": "0.684",
        "l1GasPrice": "0xb5fd948f9",
        "l1GasUsed": "0x7f0",
        "logs": [],
        "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
        "status": "0x1",
        "to": "0xe60ea7b0725e252ce2ad9f8e79eb3e686f5d1148",
        "transactionHash": "0x84e128b7c54d22f5b283538553cc0a9d9c8056c35a1010d64558dd0d86fc1fc3",
        "transactionIndex": "0x4",
        "type": "0x2"
    },
    "id": 1
}
```

So yes it's the same tx

**Reply by ajsutton** [reply]
Yes as mentioned above Etherscan is reporting the L1 Gas Used as L1GasUsed * L1FeeScalar.

**Comment by sbvegan** [comment]
@jonathanykh Etherscan just implemented a fix for this! Thank you for attention to detail and improving our ecosystem.

**Reply by jonathanykh** [reply]
Awesome! Happy to hear it and thank you for solving this issue so swiftly.
